File: server/lib/sensehat.js
/** Imports */
  import path from "path"
  import pyshell from "python-shell"

/** Bindings */
  export default class SenseHat {

    /** Rotation */
      async rotation(r, {redraw = true} = {}) {
        r = (r+360)%360
        if (![0, 90, 180, 270].includes(r))
          throw new Error(`Invalid rotation ${r}`)
        return await this.call({func:"set_rotation", r, redraw})
      }

    /** Flip */
      async flip(o, {redraw = true} = {}) {
        o = {"h":"h", "v":"v", "horizontal":"h", "vertical":"v"}[o]
        if (!o)
          throw new Error(`Invalid orientation <${o}>`)
        return await this.call({func:`flip_${o}`, redraw})
      }

    /** Set pixels */
      async pixels(pixels = null) {
        if (arguments.length)
          return await this.call({func:`set_pixels`, pixels})
        return await this.call({func:`get_pixels`})
      }

    /** Set pixel */
      async pixel({x, y}, {r, g, b, colour} = {}) {
        if (arguments.length > 1)
          return await this.call({func:`set_pixel`, x, y, r, g, b, pixel:colour})
        return await this.call({func:`get_pixel`, x, y})
      }

    /** Draw image */
      async draw(image, {redraw = true} = {}) {
        image = path.relative(path.dirname(SenseHat.py), image)
        return await this.call({func:`load_image`, file_path:image, redraw})
      }

    /** Clear */
      async clear({r, g, b, colour} = {}) {
        return await this.call({func:`clear`, r, g, b, colour})
      }

    /** Show string  */
      async string(text, {speed, colour, background}) {
        return await this.call({func:`show_message`, text_string:text, scroll_speed:speed, text_colour:colour, back_colour:background})
      }

    /** Show letter  */
      async letter(text, {colour, background}) {
        return await this.call({func:`show_letter`, s:text, text_colour:colour, back_colour:background})
      }

    /** Get humidity */
      async humidity() {
        return await this.call({func:`get_humidity`})
      }

    /** Get pressure */
      async pressure() {
        return await this.call({func:`get_pressure`})
      }

    /** Get temperature */
      async temperature({from = "humidity"} = {}) {
        let func = null
        if (["h", "humidity"].includes(from))
          func = "get_temperature_from_humidity"
        if (["p", "pressure"].includes(from))
          func = "get_temperature_from_pressure"
        if (!func)
          throw new Error(`Invalid source ${from}`)
        return await this.call({func})
      }

    /** IMU configuration */
      async imu({compass = false, gyroscope = false, accelerometer = false} = {}) {
        return await this.call({func:"set_imu_config", compass_enabled:compass, gyro_enabled:gyroscope, accel_enabled:accelerometer})
      }

    /** Orientation */
      async orientation({unit = "d"} = {}) {
        let func = null
        if (["d", "deg", "degree", "degrees"].includes(unit))
          func = "get_orientation_degrees"
        if (["r", "rad", "radian", "radians"].includes(unit))
          func = "get_orientation_radians"
        if (!func)
          throw new Error(`Invalid unit ${unit}`)
        return await this.call({func})
      }

    /** Compass */
      async compass({raw = false} = {}) {
        return await this.call({func:`get_compass${raw ? "_raw" : ""}`})
      }

    /** Gyroscope */
      async gyroscope({raw = false} = {}) {
        return await this.call({func:`get_gyroscope${raw ? "_raw" : ""}`})
      }

    /** Accelerometer */
      async accelerometer({raw = false} = {}) {
        return await this.call({func:`get_accelerometer${raw ? "_raw" : ""}`})
      }

    /** Constructor */
      constructor() {
        this.python = new pyshell.PythonShell(SenseHat.py)
        //Result and messages handlers
          this.python.on("message", message => {
            console.log(`>>> ${message}`)
            try {
              const {uid = null, error = null, result = null} = JSON.parse(message)
              this.calls.get(uid)?.[!error ? "solve" : "reject"]?.(result)
            }
            catch (error) {
              console.error(error)
            }
          })
          this.python.on("stderr", message => console.error(`SenseHat stderr >>> ${message}`))
      }

    /** Close */
      close() {
        this.python.end()
      }

    /** Calls map */
      calls = new Map()

    /** Call a function */
      async call({func, ...args}) {
        //Generate uid
          const uid = Math.random()
        //Create promise and call python
          return await new Promise((solve, reject) => {
            this.calls.set(uid, {solve, reject})
            this.python.send(JSON.stringify({uid, func, ...args}))
          })
      }

    /** Python binding file */
      static py = path.join(path.dirname(import.meta.url.replace(/file:[/]{2}/, "")), "sensehat.py")

  }

File: server/lib/sensehat.py
# This is a python interface based on https://pythonhosted.org/sense-hat/api/
# It allows mapping with NodeJS without the need of rewriting the whole library for conveniency

# Imports
from sense_hat import SenseHat, ACTION_PRESSED, ACTION_HELD, ACTION_RELEASED
import os, sys, json

# Sense hat
sensehat = SenseHat()

def sense_dump(ledmatrix = False):
  return {
    "pixels":sensehat.get_pixels() if ledmatrix else None,
    "lowlight":sensehat.low_light,
    "gamma":sensehat.gamma if ledmatrix else None,
    "humidity":sensehat.get_humidity(),
    "pressure":sensehat.get_pressure(),
    "temperature":{
      "humidity":sensehat.get_temperature_from_humidity(),
      "pressure":sensehat.get_temperature_from_pressure(),
    },
    "orientation":{
      "degrees":sensehat.get_orientation_degrees(),
      "radians":sensehat.get_orientation_radians(),
    },
    "compass":{
      "value":sensehat.get_compass(),
      "raw":sensehat.get_compass_raw(),
    },
    "gyroscope":{
      "value":sensehat.get_gyroscope(),
      "raw":sensehat.get_gyroscope_raw(),
    },
    "accelerometer":{
      "value":sensehat.get_accelerometer(),
      "raw":sensehat.get_accelerometer_raw(),
    },
  }

def configure_ledmatrix(low_light = None, gamma = None):
  if low_light is not None:
    sensehat.low_light = low_light
  if gamma is not None:
    sensehat.gamma = gamma
  return {"low_light":sensehat.low_light, "gamma":sensehat.gamma}

def get_events():
  return sensehat.stick.get_events()

def wait_for_event(emptybuffer = None):
  return sensehat.stick.wait_for_event(emptybuffer)

def joystick_event(event):
  print(json.dumps({"event":"joystick", "result":event}))
  sys.stdout.flush()
sensehat.stick.direction_any = joystick_event

def set_pixel(x, y, r = 255, g = 255, b = 255):
  return sensehat.set_pixel(x, y, r, g, b)

def set_pixels(pixels):
  return sensehat.set_pixels(pixels)

# Main loop
stdin=' '
while stdin:
  # Read line
  sys.stdout.flush()
  sys.stderr.flush()
  stdin = sys.stdin.readline()
  # Load json
  call = uid = func = context = None
  try:
    call = json.loads(stdin)
    print(json.dumps({"debug":call}))
    uid = call.pop("uid")
    func = call.pop("func")
    context = call.pop("context", "sensehat")
  except Exception as error:
    print(json.dumps({"uid":uid, "error":str(error)}))
    continue
  # Call required function
  try:
    if context == "locals":
      print(json.dumps({"uid":uid, "error":None, "result":(locals()[func](**call))}))
    elif context == "sensehat":
      print(json.dumps({"uid":uid, "error":None, "result":(getattr(sensehat, func)(**call))}))
    else:
      print(json.dumps({"uid":uid, "error":"Invalid context {}".format(context), "result":None}))
  except Exception as error:
    print(json.dumps({"uid":uid, "error":str(error), "result":None}))
